crates/net/network/src/transactions/mod.rs

```diff
@@ -576,16 +581,45 @@ where
             }
         }
 
-        // filter out known hashes, those txns have already been successfully fetched
-        self.pool.retain_unknown(&mut hashes);
+        // filter out known hashes, those txns have already been successfully fetched. most hashes
+        // will be filtered out here since this the mempool protocol is a gossip protocol, healthy
+        // peers will send many of the same hashes
+        self.pool.retain_unknown(hashes);
         if hashes.is_empty() {
             // nothing to request
             return
         }
-        // filter out already seen unknown hashes, for those hashes add the peer as fallback
-        let peers = &self.peers;
-        self.transaction_fetcher
-            .filter_unseen_hashes(&mut hashes, peer_id, |peer_id| peers.contains_key(&peer_id));
+
+        // update sizes list to match hashes if is an eth 68 message (otherwise empty
+        // `eth68_sizes` is an empty vec)
+        eth68_sizes.retain(|(hash, _)| hashes.contains(hash));
+
+        // filter out invalid entries
+        if let Some(eth68_msg) = msg.as_eth68_mut() {
```

Comment:
hmm, this looks problematic

because we're messing with the eth68 hash vec above but we now use the eth68 message here

Reply:
I don't like the way it's done either, borrowing hashes as mutable from the message, and then later destructing it. tried to only change what was necessary of the original flow though for starters. but check this out https://github.com/paradigmxyz/reth/pull/6222#discussion_r1468925323 when we have the new types in that linked issue we can destruct the message among first things in that method. atm I'm making sure that the sizes map is kept up to date with the hashes list. here https://github.com/paradigmxyz/reth/blob/652a5534a372c671eb023965e686b542c3e4ef02/crates/net/network/src/transactions/mod.rs#L584-L595 and here https://github.com/paradigmxyz/reth/blob/652a5534a372c671eb023965e686b542c3e4ef02/crates/net/network/src/transactions/mod.rs#L617-L642. can add debug asserts there though to make sure the invariant is kept.

Reply:
if we pass the `msg.as_eth68_mut()` to the validator after modifying the hashes vec (`self.pool.retain_unknown(hashes)`) then the layout is no longer consistent?

Reply:
hang on, you're right, duplicates will be slightly off the way it is now. another size value may be loaded into `unknown_hashes` than is filtered out incase of duplicate hashes.